### PR TITLE
Make sure maxMessageSize is correctly set when Thrift is configured to use a non blocking server

### LINKED ISSUE #3094: Thrift frame size errors
We are seeing the following message continuously on a tserver:
```
Read a frame size of 1458446612, which is bigger than the maximum allowable buffer sized for ALL connections.
```
I realized that #3042 reported the same thing against 2.1.  However this is against 1.10.2, and this does not appear to be because of any port scanning or other activity as such.  The reason I say that only 1 tserver is reporting this, and as soon as I drop that tserver the errors move to another tserver.  Hence it is somehow tied to a tablet.

After increasing the logging on various thing, it seem correlated with opening connections to tservers that are only handling the accumulo.metadata table but I do not have absolute proof of this.

Version 1.10.2 of accumulo
CentOS 7.5
Java Corretto-8.302.,08 (1.8.0_302-b08)
Hadoop 3.3.3 (vanilla)
Zookeeper 3.7.1

The error comes from the org.apache.thrift.server.AbstractNonblockingServer which is extended by the org.apache.accumulo.server.rpc.CustomNonBlockingServer.  The error is in the read() method of the internal FramBuffer class.

### PR DESCRIPTION
**Pull request by cshannon**
TServerUtils was not correctly setting the max frame size value on the constructor when creating a TNonblockingServerSocket leading to failures if the frame size exceeded the default.

### CODE REVIEW

**Comment by cshannon** [comment]
For the testing I went ahead and marked the integration test as SunnyDay but I can take that off if we don't want to make those tests take longer. The test seems to take about 40-45 seconds to test the max frame size against the different Thrift configurations. 

Also note I added a test for all Thrift types except for SASL as SASL is a giant pain to setup and the main issue was when non blocking server is used which is covered by the tests and is `ThriftServerType.CUSTOM_HS_HA `(default) and `ThriftServerType.THREADPOOL`. Also a test isn't really necessary anyways when using SSL or SASL as both SSL and SASL just use `ThriftUtil.transportFactory()` to create the transport factory and that utility uses our default factory which sets the length to `Integer.MAX_VALUE` and isn't configurable as it doesn't read the max message size property.

**Review by Manno15 — APPROVED** [review]
The changes look good and I did confirm this fixes the frame size error as expected. I didn't look too closely at the IT so I hope others will give a more thorough review there.

**Comment by cshannon** [comment]
> The changes look good and I did confirm this fixes the frame size error as expected. I didn't look too closely at the IT so I hope others will give a more thorough review there.

Thanks, the test is pretty simple it just sends a large mutation that is bigger than the default thrift frame size. Before the changes if you run the test you will see the tests that use a non blocking server type hang and in the tserver logs there will be thrift errors printing the max frame size message.

**Comment by cshannon** [comment]
The full IT passed except for 1 failure which was caused by #3106 

I also removed the comment from the description about this fixing #3094 was that appears to be a bit different as that is a different error and was reported against 1.10.x version which this doesn't apply to. When this is merged the commit message needs to be updated to drop that same comment.

File: test/src/main/java/org/apache/accumulo/test/functional/ThriftMaxFrameSizeIT.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import org.apache.accumulo.harness.AccumuloClusterHarness;
+import org.apache.accumulo.miniclusterImpl.MiniAccumuloConfigImpl;
+import org.apache.accumulo.server.rpc.ThriftServerType;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.thrift.TConfiguration;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+
+@Tag(SUNNY_DAY)
```
Comment (ctubbsii) [inline test/src/main/java/org/apache/accumulo/test/functional/ThriftMaxFrameSizeIT.java:38]: I wouldn't consider this a SUNNY_DAY test. We don't want to add too many of those, because we want to keep those relatively fast, compared to the full IT build.

File: server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java [hunk 2]
```diff
@@ -221,7 +223,8 @@ private static ServerAddress createThreadedSelectorServer(HostAndPort address,
       long timeBetweenThreadChecks, long maxMessageSize) throws TTransportException {
 
     final TNonblockingServerSocket transport =
-        new TNonblockingServerSocket(new InetSocketAddress(address.getHost(), address.getPort()));
+        new TNonblockingServerSocket(new InetSocketAddress(address.getHost(), address.getPort()), 0,
+            Ints.saturatedCast(maxMessageSize));
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java:227]: I didn't know about saturated cast! That's nice. What about the timeout value? You're setting that to 0? What's the consequences of that?

File: test/src/main/java/org/apache/accumulo/test/functional/ThriftMaxFrameSizeIT.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      configureForSsl(cfg,
+          getSslDir(createTestDir(this.getClass().getName() + "_" + this.testName())));
+    }
+  }
+
+  @Nested
+  class TestDefault extends TestMaxFrameSize {
+    {
+      serverType = ThriftServerType.getDefault();
+    }
```
Comment (ctubbsii) [inline test/src/main/java/org/apache/accumulo/test/functional/ThriftMaxFrameSizeIT.java:61]: Interesting strategy to use nested test classes to affect the behavior of the `@BeforeEach` method that sets up the minicluster. I think we have some other Thrift tests that could benefit from a similar usage of this. I know there's some proxy tests, at least, that I think got moved to accumulo-proxy repo... but there might be others still here.

However, I think it's a little confusing to use non-static initializer blocks to set the serverType. Can you just change those to explicit no-arg constructors instead?

**Review by ctubbsii — CHANGES_REQUESTED** [review]
I'm wondering if this is a bug in 1.10, or if it only affected us after the upgrade to Thrift 0.17. If it affects 1.10, it would be good to backport, along with the integration test. However, we're not using JUnit5 in 1.10, so we'd have to write the IT a bit differently.

File: server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java (hunk 2 above)
Comment (cshannon) [reply server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java:227]: Yeah I've been using that for a couple years, quite handy to avoid possible buffer overflows.

I'm setting it to 0 because that is the default value and what we were doing before with the previous constructor as shown here https://github.com/apache/thrift/blob/v0.17.0/lib/java/src/main/java/org/apache/thrift/transport/TNonblockingServerSocket.java#L75

The value eventually makes it's way to being set as the soTimeout() value on the socket which mean it's an infinite timeout, so basically no timeout. https://github.com/apache/thrift/blob/v0.17.0/lib/java/src/main/java/org/apache/thrift/transport/TNonblockingSocket.java#L114

We could make this configurable but for now i was just keeping the existing behavior as that would be a new config option and should be a new Issue/PR i would think.

File: test/src/main/java/org/apache/accumulo/test/functional/ThriftMaxFrameSizeIT.java (hunk 3 above)
Comment (cshannon) [reply test/src/main/java/org/apache/accumulo/test/functional/ThriftMaxFrameSizeIT.java:61]: Yeah that's definitely weird and i will definitely fix it, it was probably a by product of me messing around with this test for a long time yesterday and not thinking.

This was my first time trying to do parameterized tests with JUnit 5 and I discovered yesterday there was no way to do parameters for a constructor, JUnit 4 style, which is what I was really trying to do.

So I spent a while just trying to play around with some options so i could use parameters and have them available in the configureMiniCluster method. I was trying all kinds of things before finally coming up with the nested class idea. Anyways, at one point I was messing around with some anonymous classes and had a static initializer so probably just copied it form that and wasn't thinking.

Comment (cshannon) [reply test/src/main/java/org/apache/accumulo/test/functional/ThriftMaxFrameSizeIT.java:61]: Also in regards to your comment about it being useful for other tests, I think so too. Curiously when I was researching online about how to solve this no one really had any answers. Everyones answer was pretty much "you can't do that". I came up with the idea for the nested classes because initially I was thinking I was going to need to write a subclass for every single test which I didn't want to do and then the nested idea hit me.

**Comment by cshannon** [comment]
> I'm wondering if this is a bug in 1.10, or if it only affected us after the upgrade to Thrift 0.17. If it affects 1.10, it would be good to backport, along with the integration test. However, we're not using JUnit5 in 1.10, so we'd have to write the IT a bit differently.

Initially I thought this didn't apply but it does (my initial search didn't show the non blocking server being used). The test should be easy to convert to JUnit 4. My preference would be to keep this PR as it is for 2.1.1 so we can keep using JUnit 5 and merge it forward to main and i can create another PR for 1.10 that has the same fix but I'll fix the test so it's JUnit 4 which I'll do now.

**Comment by cshannon** [comment]
> > I'm wondering if this is a bug in 1.10, or if it only affected us after the upgrade to Thrift 0.17. If it affects 1.10, it would be good to backport, along with the integration test. However, we're not using JUnit5 in 1.10, so we'd have to write the IT a bit differently.
> 
> Initially I thought this didn't apply but it does (my initial search didn't show the non blocking server being used). The test should be easy to convert to JUnit 4. My preference would be to keep this PR as it is for 2.1.1 so we can keep using JUnit 5 and merge it forward to main and i can create another PR for 1.10 that has the same fix but I'll fix the test so it's JUnit 4 which I'll do now.

@ctubbsii - I updated based on your comments and things should be good now. I went to backport this to 1.10 and when I did I just noticed that the maxFrameSize option doesn't exist and wasn't added until this PR: https://github.com/apache/thrift/pull/2533

So this doesn't actually apply to 1.10 ....or at least if it is the same bug we can't apply this fix. I'm now wondering if the issue with #3094 is not easily fixable do to this option not being added yet. The answer might simply be to upgrade the version of thrift which supports the option and to a version where we set it which would imply to upgrade to Accumulo 2.1.1 when we release the fix.

**Review by dlmarion — APPROVED** [review]
Catching up on this, so it appears that a new constructor was added to TNonblockingServerSocket that allows us to set the frame size, is that right? Good find!

**Comment by cshannon** [comment]
> Catching up on this, so it appears that a new constructor was added to TNonblockingServerSocket that allows us to set the frame size, is that right? Good find!

Yeah that is correct, that's likely why we didn't have it initially as it didn't exist and also why we can't backport it to 1.10.x right now as the version is too old.

**Comment by ctubbsii** [comment]
I'm still a little confused about when this max message size set in the constructor would apply, as opposed to wherever we're setting it before this change. Presumably, wherever we're currently setting the max message is also needed... I don't think we were just ignoring the max message configuration option before. As I understand it, this is just additional places where we need to set it... but I don't understand the difference between these places and the old places we set it.

File: server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java (hunk 2 above)
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java:227]: I think we do have an RPC timeout configuration option. I'd be curious to know if that is being set elsewhere and needs to be set here as well. Just as with the max message size, it seems there's likely more than one place we need to set these... here, and wherever we're currently setting them before these new constructors became available to us.

**Comment by cshannon** [comment]
> I'm still a little confused about when this max message size set in the constructor would apply, as opposed to wherever we're setting it before this change. Presumably, wherever we're currently setting the max message is also needed... I don't think we were just ignoring the max message configuration option before. As I understand it, this is just additional places where we need to set it... but I don't understand the difference between these places and the old places we set it.

The issue is the non blocking server implementation. The previous way works for a blocking server but didn't work correctly for the non blocking server implementation.

File: server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java (hunk 2 above)
Comment (cshannon) [reply server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java:227]: I figured there may be a timeout option already and I could try and update this PR to include it but I didn't know how much refactoring that would take and didn't want to bloat the scope of this PR and thought it would be better as a separate issue to add in. I can create a new issue if you want so we don't forget to do it.

**Comment by cshannon** [comment]
After talking to @ctubbsii about this a bit more I'm going to do some more research before merging this to make sure it's correct and to make sure all the code paths are properly set. I'm going to look more into why it doesn't always seem to be an issue and also make sure there's nothing to do for 1.10.

**Comment by cshannon** [comment]
I took a look at this a bit today to refresh my memory after a couple weeks and at first glance it is odd that it wasn't working before as the maxMessageSize was being configured on the transport factory already. So I will dig into it more and see what's going on and why that setting doesn't work but adding the constructor param seems to fix it.

**Comment by cshannon** [comment]
@ctubbsii and @dlmarion - Ok so I finally figured out what happened. Everything was working with the maxFrameSize setting being configured on the transport factory until version 0.17.0. The test I just added passes fine without modification to TServerUtils with version 0.16.0. The reason it broke in 0.17.0 was due to the change that is part of https://github.com/apache/thrift/pull/2533 . Specifically this commit introduced the maxFrameSize option as part of the constructor https://github.com/apache/thrift/commit/66ac7b46fab85f175aec601cb48ea05408a1c186 and is necessary to configure. 

My fix works but this behavior change is obviously a breaking change and I'm not sure whether or not it's something that was intended or should maybe be addressed in Thrift itself at some point.

### TLDR version:

My commit in this PR fixes the issue and it only applies to version 2.1.0 since no other version uses Thrift 0.17.0. So this is a new regression and does not apply to previous versions, like 1.10. Also, when we configure the settings for our custom non blocking server we do still need to keep our existing maxFrameSize settings on both the transport factory configuration and the `maxReadBufferBytes` property on the options  as shown here otherwise it breaks further downstream, even if we set the constructor args. 

https://github.com/apache/accumulo/blob/706612f859d6e68891d487d624eda9ecf3fea7f9/server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java#L274-L276

### Detailed Version:

#### Why it breaks

The root cause is the changes inside of `AbstractNonblockingServer`. This commit changed how max frame size is checked during reads, here is version [0.16.0](https://github.com/apache/thrift/blob/2a93df80f27739ccabb5b885cb12a8dc7595ecdf/lib/java/src/org/apache/thrift/server/AbstractNonblockingServer.java#L355). Previously, before 0.17.0, frame size was compared against `MAX_READ_BUFFER_BYTES` which comes from `maxReadBufferBytes` and is set by us inside [TServerUtils](https://github.com/apache/accumulo/blob/706612f859d6e68891d487d624eda9ecf3fea7f9/server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java#L276)

But in version 0.17.0, this has been refactored now and instead of reading the value from `MAX_READ_BUFFER_BYTES`, this is read from `trans_.getMaxFrameSize()` as shown here in version [0.17.0](https://github.com/apache/thrift/blob/4d493e867b349f3475203ef9848353b315203c51/lib/java/src/main/java/org/apache/thrift/server/AbstractNonblockingServer.java#L335). This value is [configured](https://github.com/apache/thrift/blob/4d493e867b349f3475203ef9848353b315203c51/lib/java/src/main/java/org/apache/thrift/transport/TNonblockingServerSocket.java#L137) by the new constructor argument which we were not setting so it was always the default value and caused the failure. 

#### Why our existing config doesn't work

Now, what confused me initially is that we also set the value on the configuration of the factory so I was unsure why there were two ways to configure the setting and why ours didn't work. It turns out the reason is they are different configuration objects used for different things. 

The `maxFrameSize` that is now set by the constructor argument is applied to the `trans_` transport which is passed to `FrameBuffer` as a [constructor arg ](https://github.com/apache/thrift/blob/4d493e867b349f3475203ef9848353b315203c51/lib/java/src/main/java/org/apache/thrift/server/AbstractNonblockingServer.java#L288)and configured back in `TNonblockingServerSocket` on line 137. 

However, the `maxFrameSize` set on the configuration for the factory we create is applied to different configurations and transports as the factory that we configure doesn't get used to create `trans_` transport but is used to create the `inTrans_` and `outTrans_` transport objects in FrameBuffer as shown [here](https://github.com/apache/thrift/blob/4d493e867b349f3475203ef9848353b315203c51/lib/java/src/main/java/org/apache/thrift/server/AbstractNonblockingServer.java#L295-L296).

So there are multiple transports configured in `FrameBuffer` and the constructor arg configuration applies to one of them and the factory configuration applies to the others.

#### Why we need to keep existing config as well
 
As I said above, we still need to keep the existing configuration on the transport factory. The new constructor setting will allow the read to get further but if we don't set the factory config then we get errors downstream [here](https://github.com/apache/thrift/blob/4d493e867b349f3475203ef9848353b315203c51/lib/java/src/main/java/org/apache/thrift/transport/layered/TFramedTransport.java#L141) when reading a frame.

Furthermore, I think we need to keep setting the `maxFrameSize` configuration on `options.maxReadBufferBytes` because if the frame size is larger than that value then thrift just hangs and spins forever with no error (which is even worse). This happened when I was testing not setting it. This is because it's [waiting](https://github.com/apache/thrift/blob/2a93df80f27739ccabb5b885cb12a8dc7595ecdf/lib/java/src/org/apache/thrift/server/AbstractNonblockingServer.java#L363) for more memory to free up but it won't ever be free since the frame to be read is less than the configured max frame size so it tries to read it but it is larger than the max read bytes so it never finishes.

**Comment by dlmarion** [comment]
Great write-up! I think this is good to merge after the RPC timeout conversation thread is resolved.

File: server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java (hunk 2 above)
Comment (cshannon) [reply server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java:227]: I forgot about that, let me go ahead and create the RPC timeout configuration as a follow on thing to investigate.

Comment (cshannon) [reply server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java:227]: I opened up #3153 to investigate the RPC issue and I took a look and it looks like we should indeed to be setting the value of that property as part of the constructor argument for client timeout. We are using the property as it is returned as part of context.getclientTimeoutInMillis() and passed to the utility class [here](https://github.com/apache/accumulo/blob/c31194740080a1f12c65aec82fe46731f9da2069/server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java#L198) but it appears we are only using it for the blocking servers but not for non blocking so it should be an easy fix to do it. I will update this PR to include that change as well.

**Comment by cshannon** [comment]
@dlmarion and @ctubbsii - I looked into the RPC timeout a bit more before making the change to apply it to the server socket to fix issue #3153 and I wanted to get some feedback before I did it. As I said in my last comment it's pretty trivial to update the code to set the timeout on the server socket. However, now I'm not sure if we want to do it and what the side effects might be.

Right now the value seems to be primarily intended to be used as a client side timeout (defaulting to 120 seconds) and not as a server side timeout. This does in fact work and is applied correctly to thrift clients. It is configured in the [ClientContext](https://github.com/apache/accumulo/blob/af6b4978d42e7a1f9102a7ec9605be51c67bb595/core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java#L226) in the constructor and set as part of the timeoutSupplier and can be accessed using a [getter](https://github.com/apache/accumulo/blob/af6b4978d42e7a1f9102a7ec9605be51c67bb595/core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java#L339). I verified with some testing that the clients time out properly when setting the value. So one potential issue I see is do we actually want to re-use this same property for a server timeout when it's been used for the client timeout and in some cases you may want to make them different.

A second issue is if a timeout is applied to the server side socket as well I am a little concerned of potential consequences and changing behavior on the client side if suddenly the server has a timeout value. If the server side does timeout then the client should timeout eventually on its side as well out as long as there's an actual timeout value set (default is 120s) and in that case I think that behavior would be fine I would think and would protect the server. 

However, I noticed there are other cases where the client sets [no timeout](https://github.com/apache/accumulo/blob/af6b4978d42e7a1f9102a7ec9605be51c67bb595/core/src/main/java/org/apache/accumulo/core/rpc/ThriftUtil.java#L110) such as for manager operations. For a case like this the client would potentially hang waiting forever if the server side timed out. I actually first was alerted to this behavior when doing some testing to verify the server config worked and was seeing the client side just hang.

So I just wanted to get people's thoughts on it before I proceeded.

**Comment by cshannon** [comment]
I'm going to go ahead and merge this and leave the RPC timeout issue (#3153) to investigate and can be a follow on PR.

**Comment by ctubbsii** [comment]
@cshannon This merge caused TServerUtilsTest to start failing. It looks like it might be an easy fix, but I haven't looked at it too closely yet.

**Comment by cshannon** [comment]
@ctubbsii - Odd, when I looked at the output in CI the errors made me immediately think it was environmental because it had to do with things like expected port numbers being wrong. Sure enough I just ran the test against both 2.1 and main branches in IntelliJ and on the command line using maven and the test passes for me. I even configured InteillJ to run the test in a loop 100 times and it didn't break. So I'm not sure what is up but I can try and take a closer look tomorrow.

**Comment by ctubbsii** [comment]
> @ctubbsii - Odd, when I looked at the output in CI the errors made me immediately think it was environmental because it had to do with things like expected port numbers being wrong.

It could just be a coincidence then, perhaps related to a change in my build environment due to a system update. I can also investigate further.

**Comment by ctubbsii** [comment]
Odd. I can't reproduce this locally. Never mind, then. If I see it again, I'll consider opening a new ticket.
